modules/thermalmanager.c

```diff
@@ -52,545 +156,995 @@
+ */
+bool
+thermal_manager_object_is_registered(thermal_object_t *thermal_object)
 {
-  thermal_object_t* thermal_object = object;
-
-  send_temperature_request(thermal_object);
-
-  // set up heartbeat service to poll the temperature of the object
-  const thermal_status_configuration_t* conf =
-      &thermal_object->conf->state[thermal_object->status];
-
-  DSM_MSGTYPE_WAIT msg = DSME_MSG_INIT(DSM_MSGTYPE_WAIT);
-  /* If thermal status is in transition and we are taking several
-   * readings before change, then use speed up values for
-   * next readings. Otherwise use configured values
-   */
-  if (thermal_object->status_change_count) {
-      msg.req.mintime = 2;
-      msg.req.maxtime = 5;
-  } else {
-      msg.req.mintime = conf->mintime;
-      msg.req.maxtime = conf->maxtime;
-  }
-  msg.req.pid     = 0;
-  msg.data        = thermal_object;
-
-  broadcast_internally(&msg);
-}
+    bool is_registered = false;
 
+    if( !thermal_object )
+        goto EXIT;
 
-void dsme_register_thermal_object(thermal_object_t* thermal_object)
+    if( g_slist_find(thermal_objects, thermal_object) )
+        is_registered = true;
+
+EXIT:
+    return is_registered;
+}
+
+/** Get sensor status and temperature
+ *
+ * The status reported is the worst deviation from normal status found in
+ * the thermal objects that match the sensor_name given.
+ *
+ * For example, if there are sensors "core0", "core1", "core2" and "core3"
+ *
+ * Requesting "core0" status returns info from the "core0" as is.
+ *
+ * But requesting "core" status depends on overall status of all four
+ * matching sensors, which is highest state and temperature found, unless
+ * no sensors are in grave overheat state and at least one sensor is in
+ * low temperature states - in which case lowest state/temperature is
+ * returned.
+ *
+ * Notes:
+ *  - The status and temperature values are not modified if no matching
+ *    sensors are found
+ *  - Values cached at thermal object level are used - the temperature
+ *    is always the latest seen, but during lower level status transition
+ *    the previous (stable) status is used
+ *
+ * @param sensor_name  sensor name / sensor group name prefix
+ * @param status       where to store sensor thermal status
+ * @param temperature  where to store sensor temperature value
+ *
+ * @param return true if matching sensors were found and output values set;
+ *               false otherwise
+ */
+bool
+thermal_manager_get_sensor_status(const char *sensor_name,
+                                  THERMAL_STATUS *status, int *temperature)
 {
-    dsme_log(LOG_DEBUG,
-             "thermalmanager: %s (%s)", __FUNCTION__,
-             thermal_object->conf->name);
+    static volatile int recursing = 0;
+
+    bool ack = false;
+
+    /* In theory backend config parser should eliminate
+     * cyclic sensor dependencies - but if it still happens,
+     * it must not end up in infinite recursion  ... */
+    if( ++recursing != 1 )
```

Comment:
DSME is multithread & multiprocess kind of a thing, but this functionality is called only from the thread that is running the mainloop -> in this context it is effectively single threaded.

It is not possible to define cyclic dependencies while only the generic plugin is in use / only it provides sensor aliasing - the whole cycle just gets rejected and ignored. The check is just "dsme should never crash and burn" future proofing with paranoia hat firmly in place.

There is a bit too much list traversal there though. It could be avoided by calculating dependency trees at sensor register/unregister. But since I thought about that a bit too late and it does not really matter as long as there are only handful of sensors -> left as future improvement opportunity.